README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fgdr <img src="man/figures/logo.png" align="right" width="120px" />

(Sorry, English version of README is not availavle for now.)

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/fgdr)](https://cran.r-project.org/package=fgdr)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/fgdr?color=FF5254)](https://cran.r-project.org/package=fgdr)
[![minimal R
version](https://img.shields.io/badge/R%3E%3D-3.3.0-blue.svg)](https://cran.r-project.org/)

[![Travis build
status](https://travis-ci.org/uribo/fgdr.svg?branch=master)](https://travis-ci.org/uribo/fgdr)
[![Codecov test
coverage](https://codecov.io/gh/uribo/fgdr/branch/master/graph/badge.svg)](https://app.codecov.io/gh/uribo/fgdr?branch=master)

このパッケージは、国土地理院 基盤地図情報 (**FGD**: Fundermental
Geographic Data <https://fgd.gsi.go.jp/download/menu.php>)
からダウンロードしたデータ(バージョン4.1に対応)をR上で扱うためにデータの変換を行う関数を提供します。

## Installation

パッケージのインストールは、CRANから行います。

``` r
install.packages("fgdr")
```

開発版を利用する場合はGitHub経由でインストールします。

``` r
install.packages("remotes")
remotes::install_github("uribo/fgdr")
```

## Example

fgdrパッケージを使った、基盤地図情報データの読み込み方法を紹介します。

なお、fgdrパッケージでは、基盤地図情報のデータをダウンロードする関数は用意していません。読み込むデータは各自でダウンロードしてください。また、読み込んだデータの扱いについても[基盤地図情報ダウンロードサービスの利用規約](https://fgd.gsi.go.jp/download/terms.html)に従ってください。

このパッケージには、基盤地図情報データダウロードサービスが提供している

-   基本項目
-   数値標高モデル (5m, 10mメッシュ)

について、データフレーム、Rの地理空間データを扱うためのクラスである[sf](https://CRAN.R-project.org/package=sf)または[raster](https://CRAN.R-project.org/package=raster)、[stars](https://CRAN.R-project.org/package=stars)、[terra](https://CRAN.R-project.org/package=terra)として返却する関数が備わっています。

``` r
library(fgdr)
library(terra)
library(sf)
```

## 基本項目

基本項目のデータは`read_fgd()`で読み込みます。この関数は対象のファイルの種類に応じて、自動的に地物の種類を判別して返却します。例えば、行政区画代表点
(AdmPt) では下記のようにポイントデータです。

``` r
read_fgd("FG-GML-523346-AdmPt-20180701-0001.xml")
```

    #> Simple feature collection with 4 features and 8 fields
    #> Geometry type: POINT
    #> Dimension:     XY
    #> Bounding box:  xmin: 133.7835 ymin: 35.00697 xmax: 133.8736 ymax: 35.06013
    #> Geodetic CRS:  JGD2011
    #> # A tibble: 4 × 9
    #>   gml_id       type  name  adm_code life_span_from development_date org_gi_level
    #>   <chr>        <chr> <chr> <chr>    <date>         <date>                  <int>
    #> 1 K6_48162126… 郡市… 真庭… 33214    2015-05-25     2018-07-03              25000
    #> 2 K6_48180126… 町村… 美咲… 33666    2015-05-25     2018-07-03              25000
    #> 3 K6_48183126… 郡市… 津山… 33203    2015-05-25     2018-07-03              25000
    #> 4 K6_48192126… 郡市… 津山… 33203    2015-05-25     2018-07-03              25000
    #> # … with 2 more variables: visibility <chr>, geometry <POINT [°]>

水域 (WL) はライン、行政区画 (AdmArea) はポリゴンデータになります。

``` r
read_fgd("FG-GML-523346-AdmArea-20180701-0001.xml")
```

    #> Simple feature collection with 4 features and 8 fields
    #> Geometry type: POLYGON
    #> Dimension:     XY
    #> Bounding box:  xmin: 133.75 ymin: 35 xmax: 133.875 ymax: 35.08333
    #> Geodetic CRS:  JGD2011
    #> # A tibble: 4 × 9
    #>   gml_id       type  name  adm_code life_span_from development_date org_gi_level
    #>   <chr>        <chr> <chr> <chr>    <date>         <date>                  <int>
    #> 1 K4_523346_1… 町村… 美咲… 33666    2018-07-02     2018-07-03              25000
    #> 2 K4_523346_2… 郡市… 真庭… 33214    2018-07-02     2018-07-03              25000
    #> 3 K4_523346_3… 郡市… 津山… 33203    2018-07-02     2018-07-03              25000
    #> 4 K4_523346_4… 郡市… 津山… 33203    2018-07-02     2018-07-03              25000
    #> # … with 2 more variables: visibility <chr>, geometry <POLYGON [°]>

## 数値標高モデル

標高のメッシュデータである数値標高モデル(DEM)
のデータはデータフレームまたはrasterとして読み込みます。対象のファイルが保存されているパスおよび数値標高データの種類を指定した`read_fgd_dem()`実行します。

### 5mメッシュ

``` r
read_fgd_dem("FG-GML-5135-63-00-DEM5A-20161001.xml", 
             resolution = 5,
             return_class = "data.table")
```

    #>              type     value
    #>     1: データなし    NA [m]
    #>     2: データなし    NA [m]
    #>     3: データなし    NA [m]
    #>     4: データなし    NA [m]
    #>     5: データなし    NA [m]
    #>    ---                     
    #> 33746:     内水面 -9999 [m]
    #> 33747:     内水面 -9999 [m]
    #> 33748:     内水面 -9999 [m]
    #> 33749:     内水面 -9999 [m]
    #> 33750:     内水面 -9999 [m]

デフォルトでは返り値のオブジェクトがデータフレームですが、これは引数`return_class =`によりrasterやstars、terraのいずれかを選ぶことも可能です。terraとして読み込み、可視化する例を示します。

``` r
r <- 
  read_fgd_dem("FG-GML-5135-63-00-DEM5A-20161001.xml", 
             resolution = 5,
             return_class = "terra")
```

``` r
par(family = "IPAexGothic")
plot(r)
title(main = "数値標高モデル (5mメッシュ): 51356300",
      sub = "「基盤地図情報 数値標高モデル 5mメッシュ」（国土地理院）（https://fgd.gsi.go.jp/download/menu.php）\nをもとに瓜生真也(@uribo)が作成",
      cex.main = 1,
      cex.sub = 0.45,
      adj = 1)
```

<img src="man/figures/README-read_fgd_dem_raster_plot-1.png" width="100%" />

### 10mメッシュ

同様に10mメッシュのデータを読み込むには`resolution =`を10に変更してください。

``` r
read_fgd_dem("FG-GML-5440-10-dem10b-20161001.xml", 
             resolution = 10,
             return_class = "raster")
```

    #> class      : RasterLayer 
    #> dimensions : 750, 1125, 843750  (nrow, ncol, ncell)
    #> resolution : 0.0001111111, 0.00011112  (x, y)
    #> extent     : 140, 140.125, 36.08333, 36.16667  (xmin, xmax, ymin, ymax)
    #> crs        : +proj=longlat +ellps=GRS80 +no_defs 
    #> source     : memory
    #> names      : layer 
    #> values     : 5, 316  (min, max)

## ライセンス

MIT


NEWS.md
# fgdr (development version)

# fgdr 1.1.1

# fgdr 1.1.0

* Speed up `read_fgd_dem()` by using data.table for backend.
* Added support for terra format output with `read_fgd_dem()` 
* When a DEM file is read and returned as a df, a unit (meter) attribute is added to the elevation value.

## Bug fix

- Fix a XML element name discrepancy. ([#21](https://github.com/uribo/fgdr/issues/21)).

# fgdr 1.0.1

* Compatibility with dependent packages.

# fgdr 1.0.0

* `read_fdg()` and `read_fgd_dem()` always return JGD2011 (SRID: 6668).
* `read_fdg()` always return `sf` object regardless of what kind of file is given ([#7](https://github.com/uribo/fgdr/pull/7)).
* Change handling no data. It as `NA_real` instead of `-9999` ([#5](https://github.com/uribo/fgdr/issues/5)).
* `read_fgd_dem()`'s handling to fill up missing value cell improvement (@nonsabotage, [#3](https://github.com/uribo/fgdr/issues/3)).
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(fgdr)

test_check("fgdr")


tests/testthat/test-util.R
context("test-util")

test_that("xml schema validation", {
  expect_warning(expect_true(xml_validate(
    x = read_xml(
      system.file("extdata/FG-GML-0000-00-00-DEM5A-dummy.xml",
                  package = "fgdr")
    ),
    schema = read_xml(system.file("extdata/FGD_GMLSchema.xsd",
                                  package = "fgdr"))
  )))
  expect_warning(expect_true(xml_validate(
    x = read_xml(
      system.file("extdata/FG-GML-0000-10-dem10b-dummy.xml",
                  package = "fgdr")
    ),
    schema = read_xml(system.file("extdata/FGD_GMLSchema.xsd",
                                  package = "fgdr"))
  )))
  expect_warning(expect_true(xml_validate(
    x = read_xml(
      system.file("extdata/FG-GML-0000-10-dem10b-dummy.xml",
                  package = "fgdr")
    ),
    schema = read_xml(system.file("extdata/FGD_GMLSchema.xsd",
                                  package = "fgdr"))
  )))
  suppressWarnings(expect_true(xml_validate(
    x = read_xml(
      system.file("extdata/FG-GML-000000-AdmPt-dummy.xml",
                  package = "fgdr")
    ),
    schema = read_xml(system.file("extdata/FGD_GMLSchema.xsd",
                                  package = "fgdr"))
  )))
})

test_that("dem validation", {
  expect_equal(dem_check(
    system.file("extdata/FG-GML-0000-00-00-DEM5A-dummy.xml",
                package = "fgdr"),
    .verbose = TRUE,
    options = "NOBLANKS"
  ),
  c(0L, 0L))
  res <-
    fgd_dem_file_info(system.file("extdata/FG-GML-0000-00-00-DEM5A-dummy.xml",
                                  package = "fgdr"))
  expect_is(res, "list")
  expect_length(res, 6)
  expect_named(res, c("xml_docs", "type", "is5m", "skip_n", "limit_n", "meshcode"))
  expect_equal(res$type, "DEM")
  expect_true(res$is5m)
  expect_equal(res$meshcode, "54400098")
  res <-
    fgd_dem_file_info(system.file("extdata/FG-GML-0000-10-dem10b-dummy.xml",
                                  package = "fgdr"))
  expect_is(res, "list")
  expect_length(res, 6)
  expect_named(res, c("xml_docs", "type", "is5m", "skip_n", "limit_n", "meshcode"))
  expect_equal(res$type, "DEM")
  expect_false(res$is5m)
  expect_equal(res$meshcode, "544000")
})

test_that("standard", {
  res <-
    fgd_file_info(system.file("extdata/FG-GML-000000-AdmPt-dummy.xml",
                              package = "fgdr"))
  expect_is(res, "list")
  expect_length(res, 2)
  expect_named(res, c("xml_docs", "type"))
  expect_equal(res$type, "AdmPt")
})


tests/testthat/test-read_fgd.R
context("test-read_fgd")

test_that("Successd on dummies", {
  res <-
    read_fgd(file = system.file("extdata/FG-GML-000000-AdmPt-dummy.xml",
                                package = "fgdr"))
  expect_s3_class(res, "sf")
  expect_equal(dim(res), c(1, 9))
  expect_named(res, c("gml_id", "type", "name", "adm_code",
                      "life_span_from", "development_date",
                      "org_gi_level", "visibility", "geometry"))
})


R/util.R
#' DEM input file status check
#'
#' @param file XML file download from fgd
#' @param .verbose `logical`. suppress info input XML file's about DEM information.
#' @param ... Additional arguments passed on to other functions.
#' @import xml2
dem_check <- function(file, .verbose = TRUE, ...) {
  file_info <-
    fgd_file_info(file, ...)
  if (file_info$type != "DEM") {
    rlang::inform("Input files must be DEM format")
  } else {
    value_order <-
      file_info$xml_docs %>%
      xml2::xml_find_all("/*/*/*/gml:coverageFunction/gml:GridFunction/gml:sequenceRule") %>% # nolint
      xml2::xml_attr(attr = "order")
    if (value_order != "+x-y")
      rlang::inform("check input file format")
    start_point <-
      file_info$xml_docs %>%
      xml2::xml_find_all("/*/*/*/gml:coverageFunction/gml:GridFunction/gml:startPoint") %>% # nolint
      xml2::xml_text() %>%
      stringr::str_split("[[:space:]]") %>%
      unlist() %>%
      as.integer()
    if (.verbose == TRUE)
      if (!all.equal(start_point, c(0L, 0L)))
        rlang::inform("Data is not given from the starting point.\nCheck these coordinate")
    start_point
  }
}

set_coords <- function(raster, meshcode, as_stars = FALSE) {
  mesh <-
    meshcode %>%
    jpmesh::export_mesh() %>%
    sf::st_transform(crs = 6668)
  bb <-
    mesh %>%
    sf::st_bbox() %>%
    as.numeric()

  if (inherits(raster, "SpatRaster")) {
    terra::ext(raster) <-
      terra::ext(bb[1], bb[3], bb[2], bb[4])
    terra::crs(raster) <-
      "epsg:6668"
    return(raster)
  }
  # TODO: until stars::st_set_bbox() is released on CRAN,
  # it seems there's no means to set extent on stars objects
  # so we need to do it before converting to stars.
  raster::extent(raster) <-
    raster::extent(bb[1], bb[3], bb[2], bb[4])
  if (as_stars) {
    raster <-
      stars::st_as_stars(raster)
    sf::st_crs(raster) <-
      6668
  } else {
    # Depending on the version of PROJ, we'll see warning like
    # "Discarded datum Japanese_Geodetic_Datum_2011 in CRS definition"
    # But, as raster doesn't provide a way to specify it by EPSG code directly,
    # we just need to suppress the warning here...
    suppressWarnings(
      raster::crs(raster) <-
        sf::st_crs(6668)$proj4string
    )
  }
  raster
}

.line_parse <- function(xml_node) {
  xml_node %>%
    xml2::xml_contents() %>%
    xml2::xml_text(trim = TRUE) %>%
    stringr::str_split("\n") %>%
    purrr::map(~ stringr::str_split(.x, "[:space:]")) %>%
    purrr::map(
      ~ purrr::flatten_chr(.x) %>%
        rev() %>%
        as.numeric() %>%
        matrix(., ncol = 2, byrow = TRUE)) %>%
    purrr::map(
      ~ sf::st_linestring(.x)) %>%
    sf::st_sfc(crs = 6668)
}

extract_polygon <- function(d) {
  rowname <- NULL
  na_rows <-
    which(st_is_empty(d$geometry) == TRUE)

  if (length(na_rows) > 0) {
    d_empty <-
      d[na_rows, ]
    d_fill <-
      d[-na_rows, ] %>%
      sf::st_collection_extract("POLYGON")
    res <-
      rbind(d_empty, d_fill) %>%
      tibble::rownames_to_column()
    res$rowname <- as.numeric(res$rowname)
    res <-
      res[with(res, order(rowname)), ]
    res <-
      base::subset(res, select = -rowname)
  } else {
    res <-
      d %>%
      sf::st_collection_extract("POLYGON") %>%
      tibble::remove_rownames()
  }
  res
}

extract_xml_value <- function(x, name, name_length = 8) {
  . <- NULL
  x1 <-
    x %>%
    xml2::xml_find_all("/*/*") %>% # nolint
    purrr::set_names(seq(from = 1, to = length(.)))
  contents <-
    x %>%
    xml2::xml_find_all("/*/*/*") # nolint
  contents_name <-
    contents %>%
    xml2::xml_name()
  x_loc <-
    which(contents_name %in% name)
  x_vec <-
    contents[x_loc] %>%
    xml2::xml_contents() %>%
    as.character()
  if (length(x1) - 2 != length(x_vec)) {
    x2 <-
      seq(from = 3, to = length(x1)) %>%
      purrr::map_lgl(
        ~ x1[[.x]] %>%
          xml2::xml_length() == name_length) %>%
      purrr::set_names(seq(from = 3, to = length(x1)))
    na_loc <-
      x2 %>%
      purrr::keep(~ .x == FALSE) %>%
      names() %>%
      as.numeric()
    x_vec <-
      seq(1, length(x_vec)) %>%
      purrr::map(
        function(x) {
          tmp <- x %in% na_loc
          if (tmp == TRUE) {
            res <- c(NA_real_, x_vec[x])
          } else {
            res <- x_vec[x]
          }
          res
        }
      ) %>%
      purrr::reduce(c)
  }
  utils::type.convert(x_vec, as.is = TRUE)
}


R/commons.R
#' Line element parsed
#'
#' @inheritParams dem_check
#' @importFrom xml2 xml_find_all
#' @importFrom rlang abort
#' @import purrr
#' @details type AdmArea, BldA, WA
fgd_line_parse <- function(file) {
  file_info <- fgd_file_info(file)
  if (!file_info$type %in% c("AdmBdry", "BldL",
                             "Cntr",
                             "CommBdry",
                             "Cstline",
                             "RailCL",
                             "RdCompt", "RdEdg",
                             "WA", "WL",
                             "WStrA", "WStrL",
                             "AdmArea", "BldA")) {
    rlang::abort("input irregular type file")
  }
  if (file_info$type %in% c("AdmArea")) {
    res <-
      file_info$xml_docs %>%
      xml2::xml_find_all("/*/*/*/gml:Surface/gml:patches/gml:PolygonPatch/gml:exterior/gml:Ring/gml:curveMember/gml:Curve/gml:segments/gml:LineStringSegment/gml:posList") %>% # nolint
      .line_parse()
  }
  if (file_info$type %in% c("AdmBdry", "BldL", "Cntr", "CommBdry", "Cstline",
                            "RailCL", "RdCompt", "RdEdg", "WL", "WStrL")) {
    res <-
      file_info$xml_docs %>%
      xml2::xml_find_all("/*/*/*/gml:Curve/gml:segments/gml:LineStringSegment/gml:posList") %>% # nolint
      .line_parse()
  }

  if (file_info$type %in% c("BldA", "WA", "WStrA")) {
    res <-
      file_info$xml_docs %>%
      xml2::xml_find_all("/*/*/*/gml:Surface/gml:patches/gml:PolygonPatch/gml:exterior/gml:Ring/gml:curveMember/gml:Curve/gml:segments/gml:LineStringSegment/gml:posList") %>% # nolint
      .line_parse()
  }
  res
}

fgd_point_parse <- function(file) {
  file_info <- fgd_file_info(file)
  if (!file_info$type %in% c("AdmPt", "CommPt", "ElevPt",
                             "GCP")) {
    rlang::abort("input irregular type file")
  }
  res <-
    file_info$xml_docs %>%
    xml2::xml_find_all("/*/*/*/gml:Point/gml:pos") %>% # nolint
    xml2::xml_contents() %>%
    as.character() %>%
    purrr::map(~ stringr::str_split(.x, "[:space:]")) %>%
    purrr::flatten() %>%
    purrr::map(~ as.numeric(rev(.x)))
  res
}

elem_to_line <- function(xml_parsed) {
  xml_parsed %>%
    purrr::map(
      ~ sf::st_linestring(matrix(unlist(.x),
                                 ncol = 2,
                                 byrow = TRUE))) %>%
    sf::st_sfc(crs = 6668)

}

fgd_file_info <- function(file, ...) {
  xmls <-
    xml2::read_xml(file, ...)
  type <-
    xmls %>%
    xml2::xml_child(search = 3) %>%
    xml2::xml_name()
  list(xml_docs = xmls, type = type)
}

fgd_dem_file_info <- function(file, ...) {
  file_info <-
    fgd_file_info(file, ...)
  dem_info <-
    xml2::xml_child(file_info$xml_docs,
                    xml2::xml_children(file_info$xml_docs) %>%
                      xml2::xml_name() %>%
                      stringr::str_which("DEM"))
  is5m <-
    xml2::xml_child(dem_info,
                  dem_info %>%
                    xml2::xml_children() %>%
                    xml2::xml_name() %>%
                    stringr::str_which("type")) %>%
    xml2::xml_contents() %>%
    as.character() %>%
    stringr::str_detect("5m")
  meshcode <-
  xml2::xml_child(dem_info,
                  dem_info %>%
                    xml2::xml_children() %>%
                    xml2::xml_name() %>%
                    stringr::str_which("mesh")) %>%
  xml2::xml_contents() %>%
  as.character()
  list(xml_docs = file_info$xml_docs,
       type = file_info$type,
       is5m = is5m,
       meshcode = meshcode)
}


R/read_fdg_dem.R
#' @title Read and Parse Fundamental Geospatial Data (FGD) dem file
#'
#' @description The JPGIS (GML) format file provided by FGD as input,
#' the digital elevation models in the file are read as a data.frame or
#' spatial object (raster, stars or terra).
#' Supporting FGD Version 4.1 (2016/10/31)
#' @seealso \url{https://fgd.gsi.go.jp/download/ref_dem.html}
#'
#' @inheritParams read_fgd
#' @param resolution the number of dem mesh size resolution: 5m or 10m
#' @param return_class one of return object class: '[data.table][data.table::data.table]'
#' for faster than data.frame, '[data.frame][data.frame]', '[raster][raster::raster]',
#' '[stars][stars::st_as_stars]' or '[terra][terra::rast]'
#' @import xml2
#' @importFrom data.table fread
#' @importFrom purrr pluck reduce
#' @importFrom raster raster
#' @importFrom rlang arg_match warn
#' @importFrom tibble add_row
#' @importFrom readr read_csv
#' @importFrom terra rast
#' @importFrom stars st_as_stars
#' @return A [tibble][tibble::tibble] (data.frame), [raster][raster::raster],
#' [stars][stars::st_as_stars] or [terra][terra::rast]
#' @export
#' @examples
#' fgd_5dem <- system.file("extdata/FG-GML-0000-00-00-DEM5A-dummy.xml", package = "fgdr")
#' read_fgd_dem(fgd_5dem,
#'              resolution = 5,
#'              return_class = "data.table")
#' read_fgd_dem(fgd_5dem,
#'              resolution = 5,
#'              return_class = "data.frame")
#' # return as raster
#' read_fgd_dem(fgd_5dem,
#'              resolution = 5,
#'              return_class = "raster")
#' # return as stars
#' fgd_10dem <- system.file("extdata/FG-GML-0000-10-dem10b-dummy.xml", package = "fgdr")
#' read_fgd_dem(fgd_10dem,
#'              resolution = 10,
#'              return_class = "stars")
read_fgd_dem <- function(file, resolution = c(5, 10),
                         return_class) {
  . <- value <- NULL
  output_type <-
    rlang::arg_match(
      return_class,
      c("data.table", "data.frame", "raster", "stars", "terra"))
  if (resolution == 5) {
    grid_size <- list(x = 225, y = 150)
    xml_opts <- "NOBLANKS"
  } else if (resolution == 10) {
    grid_size <- list(x = 1125, y = 750)
    xml_opts <- "HUGE"
  }
  checked <-
    dem_check(file, .verbose = FALSE, options = xml_opts)
  file_info <-
    fgd_dem_file_info(file, options = xml_opts)
  input_x <-
    file_info$xml_docs %>%
    xml2::xml_find_all("/*/*/*/gml:rangeSet/gml:DataBlock/gml:tupleList") %>% # nolint
    xml2::xml_contents() %>%
    as.character()
  if (output_type == "data.table") {
    df_dem <-
      input_x %>%
      data.table::fread(col.names = c("type", "value"),
                        colClasses = c("character", "double"))
  } else {
    df_dem <-
      input_x %>%
      readr::read_csv(col_names = c("type", "value"),
                      col_types = c("cd"))
  }
  df_dem$value[df_dem$type == "\u30c7\u30fc\u30bf\u306a\u3057"] <- NA_real_
  if (identical(checked, c(0L, 0L))) {
    df_dem_full <-
      df_dem
  } else {
    df_dem_full <-
      df_dem %>%
      tibble::add_row(
        type = rep("\u30c7\u30fc\u30bf\u306a\u3057",
                   times = (as.numeric(checked[1]) + as.numeric(checked[2]) * grid_size$x)), # nolint
        value = NA_real_,
        .before = 0)
  }
  if (nrow(df_dem_full) < purrr::reduce(grid_size, `*`)) {
    df_dem_full <-
      df_dem_full %>%
      tibble::add_row(
        type = rep("\u30c7\u30fc\u30bf\u306a\u3057",
                   times = (purrr::reduce(grid_size, `*`) - nrow(.))),
        value = NA_real_)
  }
  if (output_type %in% c("data.table", "data.frame")) {
    res <-
      df_dem_full %>%
      purrr::modify_at(2,
                       ~ units::set_units(.x, "m"))
    return(res)
  }

  rast_mat <-
    df_dem_full %>%
    purrr::pluck("value") %>%
    matrix(nrow = grid_size$x, ncol = grid_size$y) %>%
    t()
  if (output_type == "terra") {
    res <-
      rast_mat %>%
      terra::rast() %>%
      set_coords(meshcode = file_info$meshcode)
    return(res)
  }
  # output_type is "raster" or "stars"
  res <-
    rast_mat %>%
    raster::raster()
  res %>%
    set_coords(meshcode = file_info$meshcode,
               as_stars = identical(output_type, "stars"))
}


data-raw/fgd-gml_schema.R
#####################################
# XML Schema
# http://www.data.go.jp/data/en/dataset/mlit_20140919_3102/resource/d348b555-b57f-4e1f-ae0c-51d1d028e378
#####################################
if (dir.exists("data-raw") == FALSE)
  dir.create("data-raw")

dl_url <-
  "https://fgd.gsi.go.jp/otherdata/spec/FGD_GMLSchemaV4.1.zip"

dl_file <-
  paste0("data-raw/", basename(dl_url))

path2xsd <-
  "data-raw/FGD_GMLSchemaV4.1/FGD_GMLSchema.xsd"

if (file.exists(path2xsd) == FALSE) {

  curl::curl_download(
    url = dl_url,
    destfile = dl_file
  )

  unzip(dl_file, exdir = "data-raw")
  unlink(dl_file)
}

# Modify xsd --------------------------------------------------------------
if (file.exists("test/testthat/FGD_GMLSchema.xsd") == FALSE) {
  str <- readLines(path2xsd)

  str <- sub('xmlns="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"',
             'xmlns:default="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"', str)

  cat(paste(str, collapse = "\n"),
      file = paste0("test/testthat/", basename(path2xsd)),
      append = FALSE)
}

# Create dummy XML --------------------------------------------------------
set.seed(123)
library(xml2)
library(magrittr)
library(zeallot)

# 1/2 DEM -------------------------------------------------------------
c(str_dummry_dem5, str_dummry_dem10) %<-%
  purrr::map(
    c("data-raw/FG-GML-5135-63-DEM5A/FG-GML-5135-63-00-DEM5A-20161001.xml",
      "data-raw/FG-GML-5440-10-dem10b-20161001.xml"),
    ~ sub('xmlns="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"',
          'xmlns:default="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"',
          readLines(.x))
  )

dem_types <-
  c("データなし", "内水面", "地表面", "海水面", "その他")

# 10m ---------------------------------------------------------------------
docs_dem10 <-
  read_xml(paste0(str_dummry_dem10, collapse = "\n"), options = "HUGE")

docs_dem10 %>%
  xml_find_first("/*/gml:description") %>%
  xml_set_text("基盤地図情報メタデータ ID=dummy:00-00")

docs_dem10 %>%
  xml_find_first("/*/gml:name") %>%
  xml_set_text("基盤地図情報ダウンロードデータ（GML版dummy）")

docs_dem10 %>%
  xml_find_first("/Dataset/DEM/fid") %>%
  xml_set_text("fgoid:00-00000-0-0-00000")

docs_dem10 %>%
  xml_find_first("/Dataset/DEM/lfSpanFr/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs_dem10 %>%
  xml_find_first("/Dataset/*/devDate/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs_dem10 %>%
  xml_find_first("/Dataset/DEM/orgMDId") %>%
  xml_set_text("Dummy")

# 共通...
# xml_find_first("/*/gml:name")
# xml_find_first("/Dataset/DEM/lfSpanFr/gml:timePosition")
# xml_find_first("/Dataset/*/devDate/gml:timePosition")
# xml_find_first("/Dataset/DEM/orgMDId")
# xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:lowerCorner")
# xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:upperCorner")

docs_dem10 %>%
  xml_find_first("/Dataset/DEM/mesh") %>%
  xml_set_text("000000")

docs_dem10 %>%
  xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:lowerCorner") %>%
  xml_set_text("0.0 0.000")
docs_dem10 %>%
  xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:upperCorner") %>%
  xml_set_text("0.0 0.000")

docs_dem10 %>%
  xml2::xml_find_all("/*/*/*/gml:rangeSet/gml:DataBlock/gml:tupleList") %>%
  xml_set_text(
    paste0(
      "\n",
      paste(
        # error when size = 843750
        sample(dem_types, size = 10, replace = TRUE),
        "-9999.",
        sep = ",",
        collapse = "\n"
      ),
      "\n"
    )
  )

docs_dem10 %>%
  xml_find_all("/Dataset/DEM/coverage/gml:coverageFunction/gml:GridFunction/gml:startPoint") %>%
  xml_set_text("0 0")

cat(paste(as.character(docs_dem10), collapse = "\n"),
    file = paste0("test/testthat/", "FG-GML-0000-10-dem10b-dummy.xml"),
    append = FALSE)


# 5m ----------------------------------------------------------------------
docs <-
  read_xml(paste0(str_dummry_dem5, collapse = "\n"))

docs %>%
  xml_find_first("/*/gml:description") %>%
  xml_set_text("基盤地図情報メタデータ ID=dummy:00-0000")

docs %>%
  xml_find_first("/*/gml:name") %>%
  xml_set_text("基盤地図情報ダウンロードデータ（GML版dummy）")

docs %>%
  xml_find_first("/Dataset/DEM/fid") %>%
  xml_set_text("fgoid:00-00000-00-00000-00000000")

docs %>%
  xml_find_first("/Dataset/DEM/lfSpanFr/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs %>%
  xml_find_first("/Dataset/*/devDate/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs %>%
  xml_find_first("/Dataset/DEM/orgMDId") %>%
  xml_set_text("Dummy")

docs %>%
  xml_find_first("/Dataset/DEM/mesh") %>%
  xml_set_text("00000000")

docs %>%
  xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:lowerCorner") %>%
  xml_set_text("0.0 0.000")
docs %>%
  xml_find_first("/Dataset/DEM/coverage/gml:boundedBy/gml:Envelope/gml:upperCorner") %>%
  xml_set_text("0.0 0.000")

docs %>%
  xml_find_first("/Dataset/DEM/coverage/gml:rangeSet/gml:DataBlock/gml:tupleList") %>%
  xml_set_text(
    paste0(
      "\n",
      paste(
        # size = 33750
        sample(dem_types, size = 10, replace = TRUE),
        "-9999.",
        sep = ",",
        collapse = "\n"
      ),
      "\n"
    )
  )

docs %>%
  xml_find_all("/Dataset/DEM/coverage/gml:coverageFunction/gml:GridFunction/gml:startPoint") %>%
  xml_set_text("0 0")

cat(paste(as.character(docs), collapse = "\n"),
    file = paste0("test/testthat/", "FG-GML-0000-00-00-DEM5A-dummy.xml"),
    append = FALSE)

# 2/2 Standard ---------------------------------------------------------------------
c(str_dummry) %<-%
  purrr::map(
    c("data-raw/PackDLMap/FG-GML-523345-ALL-20180701/FG-GML-523345-AdmPt-20180701-0001.xml"),
    ~ sub('xmlns="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"',
          'xmlns:default="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"',
          readLines(.x))
  )

docs <-
  read_xml(paste0(str_dummry, collapse = "\n"))

docs %>%
  xml_find_first("/*/gml:description") %>%
  xml_set_text("基盤地図情報メタデータ ID=dummy:00-0000")

docs %>%
  xml_find_first("/*/gml:name") %>%
  xml_set_text("基盤地図情報ダウンロードデータ（GML版dummy）")

docs %>%
  xml_find_first("/*/AdmPt") %>%
  xml_set_attrs(c("gml:id" = "K6_0000000000_1"))

docs %>%
  xml_find_first("/*/AdmPt/fid") %>%
  xml_set_text("00000-00000-i-00")

docs %>%
  xml_find_first("/*/AdmPt/lfSpanFr") %>%
  xml_set_attrs(c("gml:id" = "K6_0000000000_1-1"))

docs %>%
  xml_find_first("/Dataset/*/lfSpanFr/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs %>%
  xml_find_first("/*/AdmPt/devDate") %>%
  xml_set_attrs(c("gml:id" = "K6_0000000000_1-2"))

docs %>%
  xml_find_first("/Dataset/*/devDate/gml:timePosition") %>%
  xml_set_text("2018-12-01")

docs %>%
  xml_find_first("/Dataset/AdmPt/pos/gml:Point") %>%
  xml_set_attrs(c("gml:id" = "K6_0000000000_1",
                  "srsName" = "fguuid:jgd2011.bl"))

docs %>%
  xml_find_first("/Dataset/AdmPt/pos/gml:Point") %>%
  xml_set_text("36.1035 140.0850")

docs %>%
  xml_find_first("/Dataset/AdmPt/name") %>%
  xml_set_text("dummy")

docs %>%
  xml_find_first("/Dataset/AdmPt/admCode") %>%
  xml_set_text("00000")

cat(paste(as.character(docs), collapse = "\n"),
    file = paste0("test/testthat/", "FG-GML-000000-AdmPt-dummy.xml"),
    append = FALSE)

